File: java/Samples/Biometrics/VB/AbisSampleVB/Common/MainForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Gui
Imports Neurotec.IO
Imports System.Collections.Generic

Partial Public Class MainForm
	Inherits Form
	Implements ITabController
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "ITabController Members"

	Public Sub CreateNewSubjectTab(ByVal subject As NSubject) Implements ITabController.CreateNewSubjectTab
		ShowTab(GetType(EditSubjectTab), True, True, subject)
	End Sub

	Private Function OpenSubjectFunc(ByVal fileName As String, ByVal formatOwner As UShort, ByVal formatType As UShort) As NSubject
		Dim subject As NSubject = NSubject.FromFile(fileName, formatOwner, formatType)
		Dim status As NBiometricStatus = Client.CreateTemplate(subject)
		If status <> NBiometricStatus.Ok AndAlso status <> NBiometricStatus.None Then
			Utilities.ShowError("Failed to process template: {0}", status)
			Return Nothing
		End If
		Return subject
	End Function

	Public Sub OpenSubject() Implements ITabController.OpenSubject
		Using dialog = New OpenSubjectDialog()
			If dialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				Try
					Dim result As Object = LongActionDialog.ShowDialog(Me, "Opening subject", New Func(Of String, UShort, UShort, NSubject)(AddressOf OpenSubjectFunc), dialog.FileName, dialog.FormatOwner, dialog.FormatType)
					If result IsNot Nothing Then
						ShowTab(GetType(EditSubjectTab), True, True, CType(result, NSubject))
					End If
				Catch ex As Exception
					Utilities.ShowError(ex)
				End Try
			End If
		End Using
	End Sub

	Public Sub ShowSettings(ByVal ParamArray args() As Object) Implements ITabController.ShowSettings
		ShowTab(GetType(SettingsTab), False, True, args)
	End Sub

	Private Sub InitializeClient(ByVal biometricClient As NBiometricClient, ByVal clearDatabase As Boolean)
		SettingsManager.LoadSettings(biometricClient)
		biometricClient.Initialize()
		SettingsManager.LoadPreferedDevices(biometricClient)
		If clearDatabase Then
			Dim status As NBiometricStatus = biometricClient.Clear()
			If status <> NBiometricStatus.Ok Then
				Utilities.ShowInformation("Failed to clear database: {0}", status)
			End If
		End If
	End Sub

	Public Function ShowChangeDatabase() As Boolean Implements ITabController.ShowChangeDatabase
		Dim count As Integer = tabControl.TabPages.Count
		If count > 1 Then
			If (Not Utilities.ShowQuestion(Me, "Changing database will close all currently opened tabs. Do you want to continue?")) Then
				Return False
			End If

			tabControl.SelectedIndex = 0
			For i As Integer = 1 To count - 1
				tabControl.TabPages.RemoveAt(1)
			Next i
		End If

		Using dialog = New ChangeDatabaseDialog()
			If dialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				Try
					If Client IsNot Nothing Then
						Client.Dispose()
					End If
					Client = dialog.BiometricClient
					Dim operations As NBiometricOperations = Client.LocalOperations
					If (Client.RemoteConnections.Count > 0) Then
						operations = operations Or Client.RemoteConnections(0).Operations
					End If
					tsbGetSubject.Enabled = (operations And NBiometricOperations.Get) = NBiometricOperations.Get
					Return True
				Catch ex As Exception
					Utilities.ShowError(ex)
				End Try
			End If
		End Using
		Return False
	End Function

	Public Sub ShowAbout() Implements ITabController.ShowAbout
		AboutBox.Show(Me)
	End Sub

	Public Sub ShowTab(ByVal tabType As Type, ByVal alwaysCreateNew As Boolean, ByVal canClose As Boolean, ByVal ParamArray args() As Object) Implements ITabController.ShowTab
		If tabType Is Nothing OrElse (Not tabType.IsSubclassOf(GetType(TabPageContentBase))) Then
			Throw New ArgumentException("tabType")
		End If
		If (Not alwaysCreateNew) Then
			Dim tab As CloseableTabPage = tabControl.TabPages.OfType(Of CloseableTabPage)().FirstOrDefault(Function(x) x.Content IsNot Nothing AndAlso x.Content.GetType() Is tabType)
			If tab IsNot Nothing Then
				tab.Content.SetParams(args)
				tabControl.SelectedTab = tab
				Return
			End If
		End If

		Dim content As TabPageContentBase = CType(Activator.CreateInstance(tabType), TabPageContentBase)
		Dim page As CloseableTabPage = New CloseableTabPage With {.Content = content, .CanClose = canClose}
		content.Dock = DockStyle.Fill
		content.TabController = Me
		content.SetParams(args)
		page.DataBindings.Add("Text", content, "TabName")
		tabControl.TabPages.Add(page)
		tabControl.SelectedTab = page
	End Sub

	Public Sub CloseTab(ByVal tab As TabPageContentBase) Implements ITabController.CloseTab
		Dim owner As CloseableTabPage = tabControl.TabPages.OfType(Of CloseableTabPage)().First(Function(x) x.Content Is tab)

		If tabControl.TabPages.Count > tabControl.LastPageIndex Then
			tabControl.SelectedIndex = tabControl.LastPageIndex
		ElseIf tabControl.TabPages.Count > 0 Then
			tabControl.SelectedIndex = 0
		End If
		tabControl.TabPages.Remove(owner)
	End Sub

	Private privateClient As NBiometricClient
	Public Property Client() As NBiometricClient Implements ITabController.Client
		Get
			Return privateClient
		End Get
		Set(ByVal value As NBiometricClient)
			privateClient = value
		End Set
	End Property

#End Region

#Region "Private form events"

	Private Sub TsbAboutClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbAbout.Click
		ShowAbout()
	End Sub

	Private Sub TsbSettingsClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbSettings.Click
		ShowSettings()
	End Sub

	Private Sub TsbOpenSubjectClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbOpenSubject.Click
		OpenSubject()
	End Sub

	Private Sub TsbNewSubjectClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbNewSubject.Click
		CreateNewSubjectTab(New NSubject())
	End Sub

	Private Sub TbsChangeDatabaseClick(ByVal sender As Object, ByVal e As EventArgs) Handles tbsChangeDatabase.Click
		ShowChangeDatabase()
	End Sub

	Private Sub MainFormLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		ShowTab(GetType(StartTab), True, False)
	End Sub

	Private Sub MainFormFormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles MyBase.FormClosing
		tabControl.TabPages.Clear()
		If Client IsNot Nothing Then
			Client.Dispose()
			Client = Nothing
		End If
	End Sub

	Private Sub MainFormShown(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Shown
		If (Not ShowChangeDatabase()) Then
			Close()
		End If
	End Sub

	Private Sub TsbGetSubjectClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbGetSubject.Click
		Using dialog As GetSubjectDialog = New GetSubjectDialog With {.Client = Me.Client}
			If dialog.ShowDialog() = DialogResult.OK Then
				Try
					CreateNewSubjectTab(RecreateSubject(dialog.Subject))
				Catch ex As Exception
					Utilities.ShowError(ex)
				End Try
			End If
		End Using
	End Sub

#End Region

#Region "Private methods"

	Private Function RecreateSubject(ByVal subject As NSubject) As NSubject
		Dim schema As SampleDbSchema = SettingsManager.CurrentSchema
		Dim hasSchema As Boolean = Not schema.IsEmpty
		Dim galeryRecordCounts() As Integer = Nothing
		Dim resultSubject As NSubject = subject

		If hasSchema Then
			Dim bag As New NPropertyBag()
			subject.CaptureProperties(bag)

			If (Not String.IsNullOrEmpty(schema.EnrollDataName)) AndAlso bag.ContainsKey(schema.EnrollDataName) Then
				Dim templateBuffer As NBuffer = subject.GetTemplateBuffer()
				Dim enrollData As NBuffer = CType(bag(schema.EnrollDataName), NBuffer)
				resultSubject = EnrollDataSerializer.Deserialize(templateBuffer, enrollData, galeryRecordCounts)

				Dim allProperties As List(Of String) = bag.Select(Function(x) x.Key).ToList()
				Dim allowedProperties As List(Of String) = Enumerable.Union(schema.BiographicData.Elements, schema.CustomData.Elements).Select(Function(x) x.Name).ToList()
				For Each propName In allProperties.Where(Function(x) (Not allowedProperties.Contains(x)))
					bag.Remove(propName)
				Next propName

				bag.ApplyTo(resultSubject)
				resultSubject.Id = subject.Id
			End If
			If (Not String.IsNullOrEmpty(schema.GenderDataName)) AndAlso bag.ContainsKey(schema.GenderDataName) Then
				Dim genderString As String = CStr(bag(schema.GenderDataName))
				resultSubject.SetProperty(schema.GenderDataName, System.Enum.Parse(GetType(NGender), genderString))
			End If
		End If

		Return resultSubject
	End Function

#End Region

#Region "Protected methods"
	Protected Overrides Sub OnResize(ByVal e As System.EventArgs)
		MyBase.OnResize(e)

		If WindowState = FormWindowState.Maximized Then
			Update()
		End If
	End Sub
#End Region

End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/Common/MainForm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Gui;
using Neurotec.IO;

namespace Neurotec.Samples
{
	public partial class MainForm : Form, ITabController
	{
		#region Public constructor

		public MainForm()
		{
			InitializeComponent();
		}

		#endregion

		#region ITabController Members

		public void CreateNewSubjectTab(NSubject subject)
		{
			ShowTab(typeof(EditSubjectTab), true, true, subject);
		}

		public void OpenSubject()
		{
			using (var dialog = new OpenSubjectDialog())
			{
				if (dialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						NSubject result = (NSubject)LongActionDialog.ShowDialog(this, "Opening subject", new Func<string, ushort, ushort, NSubject>((fileName, formatOwner, formatType) =>
						{
							NSubject subject = NSubject.FromFile(fileName, formatOwner, formatType);
							NBiometricStatus status = Client.CreateTemplate(subject);
							if (status != NBiometricStatus.Ok && status != NBiometricStatus.None)
							{
								Utilities.ShowError("Failed to process template: {0}", status);
								return null;
							}
							return subject;
						}), dialog.FileName, dialog.FormatOwner, dialog.FormatType);

						if (result != null)
							ShowTab(typeof(EditSubjectTab), true, true, result);
					}
					catch (Exception ex)
					{
						Utilities.ShowError(ex);
					}
				}
			}
		}

		public void ShowSettings(params object[] args)
		{
			ShowTab(typeof(SettingsTab), false, true, args);
		}

		public bool ShowChangeDatabase()
		{
			int count = tabControl.TabPages.Count;
			if (count > 1)
			{
				if (!Utilities.ShowQuestion(this, "Changing database will close all currently opened tabs. Do you want to continue?"))
					return false;

				tabControl.SelectedIndex = 0;
				for (int i = 1; i < count; i++)
				{
					tabControl.TabPages.RemoveAt(1);
				}
			}

			using (var dialog = new ChangeDatabaseDialog())
			{
				if (dialog.ShowDialog() == DialogResult.OK)
				{
					if (Client != null) Client.Dispose();
					Client = dialog.BiometricClient;
					NBiometricOperations operations = Client.LocalOperations;
					if (Client.RemoteConnections.Count > 0)
						operations |= Client.RemoteConnections[0].Operations;
					tsbGetSubject.Enabled = (operations & NBiometricOperations.Get) == NBiometricOperations.Get;
					return true;
				}
			}
			return false;
		}

		public void ShowAbout()
		{
			AboutBox.Show(this);
		}

		public void ShowTab(Type tabType, bool alwaysCreateNew, bool canClose, params object[] args)
		{
			if (tabType == null || !tabType.IsSubclassOf(typeof(TabPageContentBase))) throw new ArgumentException("tabType");
			if (!alwaysCreateNew)
			{
				CloseableTabPage tab = tabControl.TabPages
					.OfType<CloseableTabPage>()
					.FirstOrDefault(x => x.Content != null && x.Content.GetType() == tabType);
				if (tab != null)
				{
					tab.Content.SetParams(args);
					tabControl.SelectedTab = tab;
					return;
				}
			}

			TabPageContentBase content = (TabPageContentBase)Activator.CreateInstance(tabType);
			CloseableTabPage page = new CloseableTabPage { Content = content, CanClose = canClose };
			content.Dock = DockStyle.Fill;
			content.TabController = this;
			content.SetParams(args);
			page.DataBindings.Add("Text", content, "TabName");
			tabControl.TabPages.Add(page);
			tabControl.SelectedTab = page;
		}

		public void CloseTab(TabPageContentBase tab)
		{
			CloseableTabPage owner = tabControl.TabPages
				.OfType<CloseableTabPage>()
				.First(x => x.Content == tab);

			if (tabControl.TabPages.Count > tabControl.LastPageIndex)
			{
				tabControl.SelectedIndex = tabControl.LastPageIndex;
			}
			else if (tabControl.TabPages.Count > 0)
			{
				tabControl.SelectedIndex = 0;
			}
			tabControl.TabPages.Remove(owner);
		}

		public NBiometricClient Client { get; set; }

		#endregion

		#region Private form events

		private void TsbAboutClick(object sender, EventArgs e)
		{
			ShowAbout();
		}

		private void TsbSettingsClick(object sender, EventArgs e)
		{
			ShowSettings();
		}

		private void TsbOpenSubjectClick(object sender, EventArgs e)
		{
			OpenSubject();
		}

		private void TsbNewSubjectClick(object sender, EventArgs e)
		{
			CreateNewSubjectTab(new NSubject());
		}

		private void TbsChangeDatabaseClick(object sender, EventArgs e)
		{
			ShowChangeDatabase();
		}

		private void TsbGetSubjectClick(object sender, EventArgs e)
		{
			using (GetSubjectDialog dialog = new GetSubjectDialog { Client = this.Client })
			{
				if (dialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						NSubject subject = RecreateSubject(dialog.Subject);
						CreateNewSubjectTab(subject);
					}
					catch (Exception ex)
					{
						Utilities.ShowError(ex);
					}
				}
			}
		}

		private void MainFormLoad(object sender, EventArgs e)
		{
			ShowTab(typeof(StartTab), true, false);
		}

		private void MainFormFormClosing(object sender, FormClosingEventArgs e)
		{
			tabControl.TabPages.Clear();
			if (Client != null)
			{
				Client.Dispose();
				Client = null;
			}
		}

		private void MainFormShown(object sender, EventArgs e)
		{
			if (!ShowChangeDatabase())
				Close();
		}

		#endregion

		#region Private methods

		private NSubject RecreateSubject(NSubject subject)
		{
			SampleDbSchema schema = SettingsManager.CurrentSchema;
			bool hasSchema = !schema.IsEmpty;
			int[] galeryRecordCounts = null;
			NSubject resultSubject = subject;

			if (hasSchema)
			{
				NPropertyBag bag = new NPropertyBag();
				subject.CaptureProperties(bag);

				if (!string.IsNullOrEmpty(schema.EnrollDataName) && bag.ContainsKey(schema.EnrollDataName))
				{
					NBuffer templateBuffer = subject.GetTemplateBuffer();
					NBuffer enrollData = (NBuffer)bag[schema.EnrollDataName];
					resultSubject = EnrollDataSerializer.Deserialize(templateBuffer, enrollData, out galeryRecordCounts);

					List<string> allProperties = bag.Select(x => x.Key).ToList();
					List<string> allowedProperties = Enumerable.Union(schema.BiographicData.Elements, schema.CustomData.Elements)
						.Select(x => x.Name)
						.ToList();
					foreach (var name in allProperties.Where(x => !allowedProperties.Contains(x)))
					{
						bag.Remove(name);
					}
					bag.ApplyTo(resultSubject);
					resultSubject.Id = subject.Id;
				}
				if (!string.IsNullOrEmpty(schema.GenderDataName) && bag.ContainsKey(schema.GenderDataName))
				{
					string genderString = (string)bag[schema.GenderDataName];
					resultSubject.SetProperty(schema.GenderDataName, Enum.Parse(typeof(NGender), genderString));
				}

			}

			return resultSubject;
		}

		#endregion

		#region Protected methods

		protected override void OnResize(EventArgs e)
		{
			base.OnResize(e);

			if (WindowState == FormWindowState.Maximized)
			{
				Update();
			}
		}

		#endregion
	}
}
